src/java.base/share/classes/java/lang/System.java

```diff
@@ -2660,6 +2662,20 @@ public StackWalker newStackWalkerInstance(Set<StackWalker.Option> options,
                                                       Continuation continuation) {
                 return StackWalker.newInstance(options, null, contScope, continuation);
             }
+
+            @Override
+            public int mismatchUTF8(String str, byte[] b, int fromIndex, int toIndex) {
+                byte[] encoded = str.isLatin1() ? str.value() : str.getBytes(UTF_8.INSTANCE);
+                if (false) {
+                    // Arrays.mismatch without the range checks (~3% faster micro getEntryHit)
+                    int aLength = encoded.length;
+                    int bLength = toIndex - fromIndex;
+                    int length = Math.min(aLength, bLength);
+                    int i = ArraysSupport.mismatch(encoded, 0, b, fromIndex, length);
+                    return (i < 0 && aLength != bLength) ? length : i;
+                }
+                return Arrays.mismatch(encoded, 0, encoded.length, b, fromIndex, toIndex);
+            }
```

Comment:
Leaving the ArraySupport.mismatch code here for now if anyone wants to investigate the ~3% regression introduced by the range checks in Arrays.mismatch

Reply:
The performance hit of using Arrays.mismatch instead of  ArraysSupport might be more like 5% actually:

```
Benchmark                             (size)  Mode  Cnt    Score    Error  Units
ZipFileGetEntry.getEntryHit              512  avgt   15   59.149 ±  0.820  ns/op
ZipFileGetEntry.getEntryHit             1024  avgt   15   73.250 ±  1.114  ns/op
ZipFileGetEntry.getEntryHitUncached      512  avgt   15  103.377 ±  1.118  ns/op
ZipFileGetEntry.getEntryHitUncached     1024  avgt   15  115.418 ±  2.767  ns/op
ZipFileGetEntry.getEntryMiss             512  avgt   15   22.200 ±  0.145  ns/op
ZipFileGetEntry.getEntryMiss            1024  avgt   15   22.528 ±  0.271  ns/op
ZipFileGetEntry.getEntryMissUncached     512  avgt   15   57.359 ±  0.428  ns/op
ZipFileGetEntry.getEntryMissUncached    1024  avgt   15   68.013 ±  2.070  ns/op
ZipFileGetEntry.getEntrySlash            512  avgt   15   72.407 ±  0.603  ns/op
ZipFileGetEntry.getEntrySlash           1024  avgt   15   82.875 ± 11.094  ns/op
```